### PR TITLE
Adding more adb commands and unit tests

### PR DESCRIPTION
**Pull request by moizjv**
- Added more adb functions and related unit tests

### CODE REVIEW

File: lib/helpers.js [hunk 1]
```diff
@@ -99,5 +93,33 @@ function getJavaForOs () {
...
+}
+
+/*
+ * Checks mScreenOnFully in dumpsys output to determine if screen is showing
+ * Default is true
+ */
+function isScreenOnFully (dumpsys) {
+  let m = /mScreenOnFully=\w+/gi.exec(dumpsys);
+  return !m || // if information is missing we assume screen is fully on
+    (m && m.length > 0 && m[0].split('=')[1] === 'true') || false;
```
Comment (jlipps) [inline lib/helpers.js]: would prefer for this second line to line up directly underneath `!m`

File: lib/tools/adb-commands.js [hunk 2]
```diff
@@ -94,4 +98,159 @@ methods.defaultIME = async function () {
...
+        canCloseKeyboard = false,
+        inputShownMatch = /mInputShown=\w+/gi.exec(stdout);
+    if (inputShownMatch && inputShownMatch[0]) {
+      isKeyboardShown = inputShownMatch[0].split('=')[1] === 'true';
+      let isInputViewShownMatch = /mIsInputViewShown=\w+/gi.exec(stdout);
+      if (isInputViewShownMatch && isInputViewShownMatch[0]) {
+        canCloseKeyboard = isInputViewShownMatch[0].split('=')[1] === 'true';
+      }
+    }
+    return {isKeyboardShown:isKeyboardShown, canCloseKeyboard:canCloseKeyboard};
```
Comment (jlipps) [inline lib/tools/adb-commands.js]: `: `

File: lib/tools/adb-commands.js [hunk 3]
```diff
@@ -94,4 +98,159 @@ methods.defaultIME = async function () {
...
+  } catch (e) {
+    log.errorAndThrow(`Error finding softkeyboard. Original error: ${e.message}`);
+  }
+};
+
+methods.sendTelnetCommand = async function (command) {
+  log.debug(`Sending telnet command to device: ${command}`);
+  let port = await this.getEmulatorPort();
+  return new Promise(function(resolve, reject) {
```
Comment (jlipps) [inline lib/tools/adb-commands.js]:
```
return new Promise((resolve, reject) => {
```

File: lib/tools/adb-commands.js [hunk 4]
```diff
@@ -94,4 +98,159 @@ methods.defaultIME = async function () {
...
+    let conn = net.createConnection(port, 'localhost'),
+        connected = false,
+        readyRegex = /^OK$/m,
+        dataStream = "",
+        res = null;
+    conn.on('connect', function () {
```
Comment (jlipps) [inline lib/tools/adb-commands.js]: `() => {`

and below a few places

File: lib/tools/adb-commands.js [hunk 5]
```diff
@@ -94,4 +98,159 @@ methods.defaultIME = async function () {
...
+        reject(new Error("Never got a response from command"));
+      } else {
+        resolve(res);
+      }
+    });
+  });
+};
+
+methods.isAirplaneModeOn = async function () {
+  let stdout = this.shell(['settings', 'get', 'global', 'airplane_mode_on']);
```
Comment (jlipps) [inline lib/tools/adb-commands.js]: don't we need to `await this.shell`? and if so, shouldn't there be a test that catches this?

File: lib/tools/adb-commands.js [hunk 6]
```diff
@@ -94,4 +98,159 @@ methods.defaultIME = async function () {
...
+  return parseInt(stdout) !== 0;
+};
+
+/*
+ * on: 1 (to turn on) or 0 (to turn off)
+ */
+methods.setAirplaneMode = async function (on) {
```
Comment (jlipps) [inline lib/tools/adb-commands.js]: in this and other methods, should we throw an error if `on` is not 1 or 0?

File: lib/tools/adb-commands.js [hunk 7]
```diff
@@ -94,4 +98,159 @@ methods.defaultIME = async function () {
...
+ */
+methods.broadcastAirplaneMode = async function (on) {
+  let args = ['am', 'broadcast', '-a', 'android.intent.action.AIRPLANE_MODE',
+              '--ez', 'state', (on === 1 ? 'true' : 'false')];
+  await this.shell(args);
+};
+
+methods.isWifiOn = async function () {
+  let stdout = await this.shell(['settings', 'get', 'global', 'wifi_on']);
+  return (parseInt(stdout) !== 0);
```
Comment (jlipps) [inline lib/tools/adb-commands.js]: `parseInt(stdout, 10)`

File: lib/tools/adb-commands.js [hunk 8]
```diff
@@ -94,4 +98,159 @@ methods.defaultIME = async function () {
...
+ * on: 1 (to turn on) or 0 (to turn off)
+ */
+methods.setWifi = async function (on) {
+  await this.shell(['am', 'start', '-n', 'io.appium.settings/.Settings', '-e',
+                    'wifi', (on === 1 ? 'on' : 'off')]);
+};
+
+methods.isDataOn = async function () {
+  let stdout = await this.shell(['settings', 'get', 'global', 'mobile_data']);
+  return (parseInt(stdout) !== 0);
```
Comment (jlipps) [inline lib/tools/adb-commands.js]: `parseInt(stdout, 10)`

**Comment by jlipps** [comment]
apart from my comments, everything looks great! :+1: for the tests

**Comment by jlipps** [comment]
travis failed due to coverage taking too long. @sebv any ideas?

**Comment by moizjv** [comment]
@jlipps made all the changes :)

File: lib/tools/adb-commands.js [hunk 9]
```diff
@@ -94,4 +98,159 @@ methods.defaultIME = async function () {
...
+  await this.shell(args);
+};
+
+methods.isWifiOn = async function () {
+  let stdout = await this.shell(['settings', 'get', 'global', 'wifi_on']);
+  return (parseInt(stdout, 10) !== 0);
+};
+
+/*
+ * on: true (to turn on) or 0 (to turn off)
```
Comment (jlipps) [inline lib/tools/adb-commands.js]: `or false (to turn off)`

File: lib/tools/adb-commands.js [hunk 10]
```diff
@@ -94,4 +98,159 @@ methods.defaultIME = async function () {
...
+ */
+methods.setDataState = async function (on) {
+  await this.shell(['am', 'start', '-n', 'io.appium.settings/.Settings', '-e',
+                    'data', on ? 'on' : 'off']);
+};
+
+/*
+ * opts: { wifi: 1/0, data 1/0 } (1 to turn on, 0 to turn off)
+ */
+methods.setWifiAndData = async function ({wifi:w, data:d}) {
```
Comment (jlipps) [inline lib/tools/adb-commands.js]: why would this be 0/1 when the others are true/false?

File: lib/tools/system-calls.js [hunk 11]
```diff
@@ -189,4 +189,30 @@ systemCallMethods.getAdbServerPort = function () {
   return process.env.ANDROID_ADB_SERVER_PORT || 5037;
 };
 
+systemCallMethods.getEmulatorPort = async function () {
+  log.debug("Getting running emulator port");
+  if (this.emulatorPort !== null) {
+    return this.emulatorPort;
+  }
+  try {
+    let devices = this.getConnectedDevices();
```
Comment (jlipps) [inline lib/tools/system-calls.js]: we don't need to `await` this?

**Pull request merged**